WordPress/gutenberg
# Components: Set up README auto-generator

Closes #58566

## What?

Sets up a system to auto-generate minimal README files for components in `@wordpress/components`.

The readme file will include the component description and prop types information from TypeScript, as well as those for any specified subcomponents. Most importantly, it includes a visible callout at the top to nudge viewers to the Storybook docs page.

## Why?

- To direct more people to the Storybook docs as the source of truth.
- Decrease maintenance cost and improve accuracy of the docs, since we currently rely on a system of manually syncing the README content with the canonical TSDocs.

Note that the ultimate goal is not to improve the README files (and thus the documentation hosted in the [Block Editor Handbook](https://developer.wordpress.org/block-editor/reference-guides/components)), but to increase awareness of the Storybook as a better source of component documentation. Therefore, we can keep the READMEs as simple as possible.

## How?

To keep this system as low-cost as possible, I propose we start with a system where a component can opt into the auto-generator with a `docs-manifest.json`. This allows us to gradually convert the existing readmes without having to build a sophisticated system upfront, with automatic handling of every edge case. We can deal with edge cases as we encounter them.

### What happens to content that currently only exists in the readme, like usage guidelines?

I'm planning to move this kind of extra content to a markdown file in the `stories` folder, so they may eventually be reviewed and reused as part of the docs in Storybook. (A lot of it is probably outdated anyway, and all crucial documentation should already be in the Storybook.)

## Testing Instructions

Try adding a `docs-manifest.json` to one of the components and run `npm run docs:components`. (If it errors, I haven't handled that edge case yet 😇 Pick something simpler!)

I have confirmed that a CI check [will error](https://github.com/WordPress/gutenberg/actions/runs/11283259631/job/31382353071?pr=66035) if someone forgets to commit a re-generated README when they should. This has the additional benefit of acting as a basic regression test for our types — we will now know if a PR causes any unexpected changes in the public API.
